File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        stack = []
        n = len(nums)
        ans = [-1] * n

        for i in range(2*n-1, -1, -1):
            while stack and stack[-1] <= nums[i%n]:
                stack.pop()
            
            #consider only indexes: 0 -> n & not 0 -> 2n becoz I only need 0->n indexes to insert in ans[].
            #After poping elements from stack using above while loop check if stack is not empty
            #If both these are true, then only add stack's top to ans[]
            if stack and i < len(nums):
                ans[i] = stack[-1]
            
            stack.append(nums[i%n])

        return ans
                    





        

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        Stack<Integer> s = new Stack<>();

        for(int i = 2*n-1; i >= 0; i--){
            int index = i % n;
            while(!s.isEmpty() && nums[index] >= nums[s.peek()]){
                s.pop();
            }

            res[index] = s.isEmpty() ? -1 : nums[s.peek()];
            s.push(index);
        }
        return res;
    }
}

